commit d10c4ec8b4bc02f3874c7ef3c3539e4e7e123969
Author: Stefanik Gábor <netrolller.3d@gmail.com>
Date:   Wed Sep 3 11:26:59 2008 +0800

    iwlwifi: enable packet injection for iwlagn
    
    Handle station IDs of transmitted packets when in monitor mode, and
    remove the various anti-injection checks from the iwl4965 driver.
    This makes injection work on iwl4965 and iwl5000. Tested on both cards.
    
    Note: To inject management frames with encryption, HW crypto support
    must be disabled using the "swcrypto=1" modparam (or "swcrypto50=1"
    for iwl5000). Otherwise most management frames won't be transmitted.
    
    Signed-off-by: Gábor Stefanik <netrolller.3d@gmail.com>
    Signed-off-by: Zhu Yi <yi.zhu@intel.com>
    Signed-off-by: John W. Linville <linville@tuxdriver.com>

diff --git a/drivers/net/wireless/iwlwifi/iwl-sta.c b/drivers/net/wireless/iwlwifi/iwl-sta.c
index 6283a3a707f5..5b7b05c8773f 100644
--- a/drivers/net/wireless/iwlwifi/iwl-sta.c
+++ b/drivers/net/wireless/iwlwifi/iwl-sta.c
@@ -968,6 +968,11 @@ int iwl_get_sta_id(struct iwl_priv *priv, struct ieee80211_hdr *hdr)
 		iwl_print_hex_dump(priv, IWL_DL_DROP, (u8 *) hdr, sizeof(*hdr));
 		return priv->hw_params.bcast_sta_id;
 
+	/* If we are in monitor mode, use BCAST. This is required for
+	 * packet injection. */
+	case IEEE80211_IF_TYPE_MNTR:
+		return priv->hw_params.bcast_sta_id;
+
 	default:
 		IWL_WARNING("Unknown mode of operation: %d\n", priv->iw_mode);
 		return priv->hw_params.bcast_sta_id;